### PR TITLE
Json Formatter Implementation

### PR DESCRIPTION
**Pull request by Mih-Mir**

### CODE REVIEW

File: CustomJsonFormatter/Startup.cs [hunk 1]
```diff
@@ -1,24 +1,31 @@
...
 {
     public class Startup
     {
-        // This method gets called by the runtime. Use this method to add services to the container.
-        // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
         public void ConfigureServices(IServiceCollection services)
         {
+            services.AddSingleton<IProfileRepository, ProfileRepository>();
+
+            var articleRepository = new ArticleRepository();
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Startup.cs:18]: why new? Just register as AddSingleton <IArticleRepository, ArticleRepository>

And it is better to rename to InMemoryArticleRepository

File: CustomJsonFormatter/Startup.cs [hunk 2]
```diff
@@ -1,24 +1,31 @@
...
+
+            services.AddSingleton<IArticleRepository>(articleRepository);
+
+            services.AddSingleton<CustomJsonFormatterService>(new CustomJsonFormatterService(articleRepository));
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Startup.cs:22]: Use provider factory. Refactor it.
Read about factory approach for this

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            _articleRepository = articleRepository;
+        }
+
+        public object GetOutputJsonObject()
+        {
+            return _articleRepository.Get();
+        }
+
+        public object GetOutputJsonObject(int id, HttpRequest request)
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Services/CustomJsonFormatterService.cs:29]: return string

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            var article = _articleRepository.GetById(id);
+
+            if (request.Headers.Values.Contains(AppJson))
+            {
+                return article;
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Services/CustomJsonFormatterService.cs:35]: Create Strategy
IArticleFormatterStrategy

string FormatArticle(Article article, FormatContext context)

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        public object GetOutputJsonObject()
+        {
+            return _articleRepository.Get();
+        }
+
+        public object GetOutputJsonObject(int id, HttpRequest request)
+        {
+            var article = _articleRepository.GetById(id);
+
+            if (request.Headers.Values.Contains(AppJson))
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Services/CustomJsonFormatterService.cs:33]: Implement factory

IArticleFormatterStrategy GetFormatter(string)

File: CustomJsonFormatter/Startup.cs (hunk 1 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Startup.cs:18]: Done

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs (hunk 5 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Services/CustomJsonFormatterService.cs:33]: Done

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs (hunk 4 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Services/CustomJsonFormatterService.cs:35]: Done

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs (hunk 3 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Services/CustomJsonFormatterService.cs:29]: Done

File: CustomJsonFormatter.ResourceAccess/Repositories/Interfaces/IInMemoryArticleRepository.cs [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 namespace CustomJsonFormatter.ResourceAccess.Repositories.Interfaces
 {
-    public interface IArticleRepository
+    public interface IInMemoryArticleRepository
```
Comment (sachenko-igor) [inline CustomJsonFormatter.ResourceAccess/Repositories/Interfaces/IInMemoryArticleRepository.cs:6]: Interface name should not be changed
Interface name IArticleRepository
Implementations: InMemoryArticleRepository, SqlArticleRepository, etc

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs [hunk 7]
```diff
@@ -12,45 +14,34 @@ public class CustomJsonFormatterService : ICustomJsonFormatterService
...
         private const string AppJsonCustom = "application/json+custom";
 
-        private const string ProfilePath = "/api/profile/";
+        private readonly IEnumerable<IArticleFormatterStrategy> _articleFormatterStrategies;
 
-        private readonly IArticleRepository _articleRepository;
+        private readonly IInMemoryArticleRepository _articleRepository;
 
-        public CustomJsonFormatterService(IArticleRepository articleRepository)
+        public CustomJsonFormatterService(IEnumerable<IArticleFormatterStrategy> articleFormatterStrategies,
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Services/CustomJsonFormatterService.cs:21]: This is a factory class, rename it. It should contain Factory word. For example ArticleFormatterStrategyFactory
Add a method "bool IsApplicable(FormatContext context)" to IArticleFormatterStrategy.
In the factory use this method to find a correct strategy.
Add the heades collection to FormatContext and use it to identify if this particular strategy is applicable.

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs [hunk 8]
```diff
@@ -12,45 +14,34 @@ public class CustomJsonFormatterService : ICustomJsonFormatterService
...
         }
 
-        public object GetOutputJsonObject()
+        public IList<Article> GetArticles()
         {
             return _articleRepository.Get();
         }
 
-        public object GetOutputJsonObject(int id, HttpRequest request)
+        public IArticleFormatterStrategy GetFormatter(ICollection<StringValues> headers)
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Services/CustomJsonFormatterService.cs:33]: change to GetFormatterStrategy(FormatterContext context)

File: CustomJsonFormatter.ResourceAccess/Repositories/Interfaces/IInMemoryArticleRepository.cs (hunk 6 above)
Comment (Mih-Mir) [reply CustomJsonFormatter.ResourceAccess/Repositories/Interfaces/IInMemoryArticleRepository.cs:6]: Ok

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs (hunk 8 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Services/CustomJsonFormatterService.cs:33]: Done

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs (hunk 7 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Services/CustomJsonFormatterService.cs:21]: Fixed

File: CustomJsonFormatter/Controllers/ArticlesController.cs [hunk 9]
```diff
@@ -8,17 +8,17 @@ namespace CustomJsonFormatter.Controllers
     [Route("api/articles")]
     public class ArticlesController : ControllerBase
     {
-        private readonly ICustomJsonFormatterService _customJsonFormatterService;
+        private readonly IArticleFormatterStrategyFactory _articleFormatterStrategyFactory;
```
Comment (sachenko-igor) [inline CustomJsonFormatter/Controllers/ArticlesController.cs:11]: Add additional layer - CustomJsonFormatterService
The responsibility of this service will be:
1. data load from the repository
2. getting a correct strategy from the strategy factory
3. executing strategy method

Controller should execute the method of the service

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs (hunk 7 above)
Comment (sachenko-igor) [reply CustomJsonFormatter/Services/CustomJsonFormatterService.cs]: Not fixed. You added this method(IsApplicable) into IFactory. It should be added to IStrategy.

File: CustomJsonFormatter/Controllers/ArticlesController.cs (hunk 9 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Controllers/ArticlesController.cs]: Done

File: CustomJsonFormatter/Services/ArticleFormatterStrategyFactory.cs [hunk 10]
```diff
@@ -46,7 +46,7 @@ public IArticleFormatterStrategy GetFormatterStrategy(FormatContext formatContex
             return null;
         }
 
-        private bool IsApplicable(FormatContext formatContext, string s)
```
[the remark this answers is not in the document]
Comment (Mih-Mir) [reply CustomJsonFormatter/Services/ArticleFormatterStrategyFactory.cs]: Done

File: CustomJsonFormatter/Services/CustomJsonFormatterService.cs (hunk 7 above)
Comment (Mih-Mir) [reply CustomJsonFormatter/Services/CustomJsonFormatterService.cs]: Fixed

**Review by sachenko-igor — APPROVED** [review]

**Pull request merged**